docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,754 @@
+ 0x2f, 0x50, 0xdc, 0xb4,                \
+ 0x79, 0x9c,                            \
+ 0x45, 0x07,                            \
+ 0xa1, 0xe9,                            \
+ 0x86, 0x2c, 0x62, 0x9b, 0x76, 0x2a}    \
+}
+
+/*! sgx_attestation_plugin
+ * 
+ * Return the SGX attesation plug-in.
+ */
+oe_attestation_plugin_t* sgx_attestation_plugin();
+
+```
+
+`sgx_attestation_plugin.c`
+
+```C
+#include "sgx_attestation_plugin.h"
+
+static
+oe_result_t
+sgx_attestation_plugin_on_register(
+    oe_attestation_plugin_t* plugin_context,
+    const void* config_data,
+    size_t config_data_size)
+{
+    OE_UNUSED(plugin_context);
+    OE_UNUSED(config_data);
+    OE_UNUSED(config_data_size);
+
+    // Nothing to do
+    return OE_OK;
+}
+
+static 
+oe_result_t 
+sgx_attestation_plugin_on_unregister(
+    oe_attestation_plugin_t* plugin_context)
+{
+    OE_UNUSED(plugin_context);
+
+    // Nothing to do
+    return OE_OK;
+}
+
+static 
+oe_result_t 
+sgx_attestation_plugin_get_evidence(
+    oe_attestation_plugin_t* plugin_context,
+    uint32_t flags,
+    const oe_claim_t* custom_claims,
+    size_t custom_claims_length,
+    const void* opt_params,
+    size_t opt_params_size,
+    uint8_t** evidence_buffer,
+    size_t* evidence_buffer_size,
+    uint8_t** endorsements_buffer,
+    size_t* endorsements_buffer_size)
+{
+    OE_UNUSED(plugin_context);
+
+    /*
+     * Pseudocode description instead of actual C code:
+     *
+     * Hash custom claims field.
+     * Call oe_get_report with the flags and opt_param parameters and the hash as reportdata.
+     * Report contains the endorsements, so extract them out.
+     * Evidence will be report + custom_claims blob.
+     */
+
+    return OE_OK;
+}
+
+static 
+oe_result_t 
+sgx_attestation_plugin_free_evidence(
+    oe_attestation_plugin_t* plugin_context,
+    uint8_t* evidence_buffer)
+{
+    OE_UNUSED(plugin_context);
+
+    return oe_free_report(evidence_buffer);
+}
+
+static 
+oe_result_t 
+sgx_attestation_plugin_free_endorsements(
+    oe_attestation_plugin_t* plugin_context,
+    uint8_t* endorsements_buffer)
+{
+    OE_UNUSED(plugin_context);
+
+    return oe_free_endorsements(endorsements_buffer);
+}
+
+static 
+oe_result_t 
+sgx_attestation_plugin_verify_evidence(
+    oe_attestation_plugin_t* plugin_context,
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size,
+    const uint8_t* endorsements_buffer,
+    size_t endorsements_buffer_size,
+    const oe_datetime_t* input_validation_time,
+    oe_claim_t** claims,
+    size_t* claims_length)
+{
+    OE_UNUSED(plugin_context);
+
+    /*
+     * Pseudocode description instead of actual C code:
+     *
+     * Call oe_verify_report will all the input parameters and get the oe_identity_t back.
+     * Look for the custom claims in the evidence header and extract them if found.
+     * Verify the hash of custom claims == report data field in evidence report.
+     * Convert oe_identity_t to the claims format.
+     */
```

Comment:
Another technical comment, which for this PR is just editorial, but may result in code changes elsewhere if needed:

Since the verify API may be running on a different machine with a different architecture than the one that generated the evidence, I think the /* comments */ should point out that the verify code needs to be smart about the byte order, so that if the SDK is compiled on a big-endian machine, it can still parse SGX evidence generated by a little-endian machine.  So one cannot, for example, simply cast to a struct and expect 32-bit int fields have the values you think they are.

I think this is important to call out here so that as people write more plugins, they don't make common mistakes.  (Either that or we have to redefine OE to not have a cross-plat goal and restrict it to only little-endian architectures, which I think runs counter to the goals of OE.)

Reply:
This is a great point!

Reply:
I added an explanation about the endianess for both the get_evidence and verify_evidence.